Arch arm cortex m vector table rework

ioannisg opened the pull request:
Ready for review.
This PR does the following changes
- removes z_arm_reserved function pointer from the Cortex-M vector table entries that correspond to exceptions reserved for future use by the ARM specification. This should be a "no-op" as these addresses are not read/written/fetched by the code. The entries are set to 0x0.
- renames z_arm_reserved to z_arm_exc_spurious and use the function pointer for vector entries that correspond to existing but not used exception
- adds a test to verify the behavior of z_arm_exc_spurious

ioannisg committed:
arch: arm: aarch32: cortex_m: write 0x0 to reserved exception entries

Write 0x0 instead of z_arm_reserved to vector exception
entries that are always reserved for future use by the
ARM architecture. These vector table entries cannot be
fetched to be executed by the Cortex-M exception entry,
so having z_arm_reserved gives a false impression, since
it is a function that may be invoked in the code. This
modification is safe since these vector entries are also
not supposed to be read / written by the code.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: aarch32: add z_arm_reserved only when we have SysTick

If the Cortex-M core does not implement the SysTick peripheral,
we should not be adding z_arm_reserved in the corresponding
vector table entry. If we do have SysTick implemented but we
are not using it as the system timer, we shall install the
reserved interrupt at the vector table entry.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: aarch32: cortex_m: add z_arm_reserved only if core has SE

If the Cortex-M core does not implement the Security Extension,
we should not be adding z_arm_reserved in the corresponding
vector table entry. That is because the entry is reserved by
the ARM architecture.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

stephanosio commented on arch/arm/core/aarch32/cortex_m/vector_table.S:
```diff
@@ -38,37 +38,45 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
     .word z_arm_usage_fault
 #if defined(CONFIG_ARM_SECURE_FIRMWARE)
     .word z_arm_secure_fault
 #else
+#if defined(CONFIG_ARMV8_M_SE)
```
Maybe we can do `#elif` here?

ioannisg replied:
Nope, I don't want to do that because it could be confused that ARMv8_M_SE and ARM_SECURE_FIRMWARE are mutually exclusive, but they are not. In fact, ARM_SECURE_FIRMWARE requires ARMV8_M_SE.

ioannisg replied:
Maybe re-organize the #ifdefs here, to have ARMV8_M_SE as external and ARM_SECURE_FIRMWARE as internal ifdef, would be better.

ioannisg replied:
done

stephanosio commented on arch/arm/core/aarch32/cortex_m/vector_table.S:
```diff
@@ -38,37 +38,45 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
 #if defined(CONFIG_SYS_CLOCK_EXISTS)
     .word z_clock_isr
 #else
     .word z_arm_reserved
```
I am not sure `z_arm_reserved` is the (semantically) correct name for it.

In this patch, we are setting all the "actually reserved" vectors to 0 and the valid-but-unused ones to `z_arm_reserved`, which is a little confusing.

ioannisg replied:
Agree - z_arm_reserved, probably was named like this long long ago, because it was meant to be written in the ARM Reserved entries. Feel free to suggest a different name.

stephanosio replied:
Maybe `z_arm_exc_spurious`, like exception-equivalent of `z_irq_spurious`?

ioannisg replied:
maybe yes, but let me think a bit more about it.
It should be out of the scope of this PR, which basically tries to address an issue with LPC5569 that @MaureenHelm  has raised.

ioannisg replied:
Changed to z_arm_exc_spurious, thanks @stephanosio

galak commented:
I wonder if we are going to make a change like this, if we should also move it out of asm code like arc has.  Look at `arch/arc/core/vector_table.c`.  I don't see any particular need for this to be asm and changing it to .c might make it be one less thing to deal with for other compilers in the future.

galak reviewed (commented):
Is z_arm_reserved used anymore after this change?

stephanosio reviewed (commented):
In general, I think setting the reserved vectors to 0 is okay; after all, we do not care about the reserved stuff and that is how reserved things are often handled (and expected to be, in case of #23972).

ioannisg commented:
> Is z_arm_reserved used anymore after this change?

Yes, it is used to trigger a CPU fault, see fault.S

ioannisg commented:
> I wonder if we are going to make a change like this, if we should also move it out of asm code like arc has. Look at `arch/arc/core/vector_table.c`. I don't see any particular need for this to be asm and changing it to .c might make it be one less thing to deal with for other compilers in the future.

No objections to that, but I would prefer to do this once for all cortex- variants together, maybe a task for @stephanosio  (?) :)

stephanosio commented:
> > I wonder if we are going to make a change like this, if we should also move it out of asm code like arc has. Look at `arch/arc/core/vector_table.c`. I don't see any particular need for this to be asm and changing it to .c might make it be one less thing to deal with for other compilers in the future.
> 
> No objections to that, but I would prefer to do this once for all cortex- variants together, maybe a task for @stephanosio (?) :)

I believe that would be feasible for Cortex-M, though not the rest (i.e. Cortex-A and Cortex-R).

There is one major difference between the vector tables for Cortex-M and the rest -- that the former consists of the **pointers** to the handlers, whereas the latter (usually) consists of the **jump instructions** to the handlers.

Moreover, in AArch64, the space allocated for each vector has increased so that you can actually put small handler implementations in the vector table itself.

ioannisg commented:
> I believe that would be feasible for Cortex-M, though not for the rest (i.e. Cortex-A and Cortex-R).

In this case I am less eager to do this for Cortex-M now - I am not 100% happy with having the cortex-m in .c while cortex-r, a in .S format...

MaureenHelm commented:
I didn't go down this path before because it's a more intrusive change, but if you're ok with it then so am I. :)

I agree with @stephanosio that `z_arm_reserved` no longer makes sense. I think it should be renamed in this PR.

ioannisg committed:
arch: arm: aarch32: rename z_arm_reserved to z_arm_exc_spurious

In the Cortex-M exception table we rename z_arm_reserved()
function to z_arm_exc_spurious(), as it is invoked when
existing (that is, non-reserved) but un-installed exceptions
are triggered, accidentaly, by software, or hardware. This
currently applies to SysTick and SecureFault exceptions.

Since fault.S is shared between Cortex-M and other AARCH32
architectures, we keep z_arm_reserved as a defined symbol
there. This commit does some additional, minor, "no-op"
cleanup in #ifdef's for Cortex-M and Cortex-R.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

stephanosio reviewed (commented):
Looks ok to me.

`z_arm_reserved` can be completely removed for Cortex-R too since it is currently not used by anything and the future Cortex-A support will use `z_arm_exc_spurious` as well for the hypervisor vector.

ioannisg commented:
> Looks ok to me.
> 
> `z_arm_reserved` can be completely removed for Cortex-R too since it is currently not used by anything and the future Cortex-A support will use `z_arm_exc_spurious` as well for the hypervisor vector.

Would you mind doing this in your clean up work for cortex-r?

stephanosio commented:
> > Looks ok to me.
> > `z_arm_reserved` can be completely removed for Cortex-R too since it is currently not used by anything and the future Cortex-A support will use `z_arm_exc_spurious` as well for the hypervisor vector.
> 
> Would you mind doing this in your clean up work for cortex-r?

Sure, that would work too.

ioannisg committed:
tests: arch: arm: interrupt: add test-case for spurious exception

We add a simple test case to verify the behavior
of z_arm_exc_spurious handler. We use the SysTick
interrupt for that so the test is enabled for
platforms that have but do not use the SysTick.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> I didn't go down this path before because it's a more intrusive change, but if you're ok with it then so am I. :)
> 
> I agree with @stephanosio that `z_arm_reserved` no longer makes sense. I think it should be renamed in this PR.

@MaureenHelm so, i think this is a good cleanup and will solve the issue with the LPC image boot. I also did the rename of z_arm_reserved in this PR.

stephanosio commented on arch/arm/core/aarch32/fault_s.S:
```diff
@@ -78,14 +83,16 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_usage_fault)
 SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_debug_monitor)
+#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_exc_spurious)
 #elif defined(CONFIG_ARMV7_R)
```
Since we are doing `#if defined(CONFIG_CPU_CORTEX_M)` above, I wonder if this should be changed to `#elif defined(CONFIG_CPU_CORTEX_R)` for consistency; after all, that should be the preferred approach since this is not ARM**v7**-specific.

ioannisg replied:
Yes, that's fine.
Done (also some more cleanup)

stephanosio commented on arch/arm/core/aarch32/fault_s.S:
```diff
@@ -30,14 +31,16 @@ GTEXT(z_arm_usage_fault)
 GTEXT(z_arm_debug_monitor)
+#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+GTEXT(z_arm_exc_spurious)
 #elif defined(CONFIG_ARMV7_R)
```
same here

ioannisg replied:
Done

stephanosio reviewed (commented):
Looks ok to me overall.

Just a minor comment about the `#if`s; though, this is really not that important since I am splitting Cortex-M and Cortex-A/-R asm exception handling code into separate files in the upcoming PR.

MaureenHelm commented on tests/arch/arm/arm_interrupt/src/arm_interrupt.c:
```diff
@@ -132,6 +141,30 @@ void test_arm_interrupt(void)
+#if defined(CONFIG_CPU_CORTEX_M_HAS_SYSTICK)
+#if !defined(CONFIG_SYS_CLOCK_EXISTS) || !defined(CONFIG_CORTEX_M_SYSTICK)
+	/* Verify that triggering a Cortex-M exception (accidentally) that has
+	 * not not been  installed in the vector table, leads to the reserved
```
typo: not not

MaureenHelm reviewed (approved):
Minor typo in the test comment, but otherwise this looks great. Thanks @ioannisg

ioannisg commented:
> Minor typo in the test comment, but otherwise this looks great. Thanks @ioannisg

I'll fix the typo, thanks for catching. Have you tested that it works for the LPC SoC, @MaureenHelm ?

MaureenHelm commented:
> Have you tested that it works for the LPC SoC, @MaureenHelm ?

yes

ioannisg commented:
> > Is z_arm_reserved used anymore after this change?
> 
> Yes, it is used to trigger a CPU fault, see fault.S
> This is needed in some cases. For example, if someone pends the SysTick IRQ without having it wired.

@galak  this is now even unit-tested :)

MaureenHelm commented:
@galak are you ok with this?
